src/Turtle/Prelude.hs

```diff
@@ -1090,6 +1093,32 @@ time io = do
 hostname :: MonadIO io => io Text
 hostname = liftIO (fmap Text.pack getHostName)
 
+-- | Show the full path of an executable file
+which :: MonadIO io => FilePath -> io (Maybe FilePath)
+which cmd = fold go Control.Foldl.head
+  where
+    go :: Shell FilePath
+    go =
+      need "PATH" >>= \case
+          Nothing -> empty
+          Just paths -> do
+            path <- select (Text.split (== ':') paths)
+            let path' = Filesystem.fromText path </> cmd
+            exists <- testfile path'
+            if exists
+                then do
+                    perms <- liftIO $
+                        getmod path'
+                          `catchIOError`
+                            \e ->
```

Comment:
That does look nicer, but it doesn't work when the action is a do block! How would you structure this?

```haskell
then do
    let handler e = ...
         action = do
             ....
             ....
    action `catcIOError` handler
```

(assuming that `let action = ...` isn't ideal)